crypto/box/src/hkdf_box.rs

```diff
@@ -0,0 +1,164 @@
+use crate::traits::{CryptoBox, Error};
+
+use aead::{
+    generic_array::{
+        sequence::{Concat, Split},
+        typenum::{Sum, Unsigned},
+        ArrayLength, GenericArray,
+    },
+    Aead, Error as AeadError, NewAead,
+};
+use core::{
+    convert::TryFrom,
+    marker::PhantomData,
+    ops::{Add, Sub},
+};
+use digest::{BlockInput, Digest, FixedOutput, Input, Reset};
+use hkdf::Hkdf;
+use mc_crypto_keys::{Kex, ReprBytes};
+use rand_core::{CryptoRng, RngCore};
+
+/// Represents a generic implementation of CryptoBox using Hkdf, a KexAlgo, and an Aead.
+///
+/// This structure contains the actual cryptographic primitive details, and
+/// specifies part of the wire format of the "footer" where the ephemeral
+/// public key comes first, and the mac comes second.
+///
+/// Preconditions:
+/// - Only stateless AEAD is supported. The build will fail if you only have AeadMut.
+pub struct HkdfBox<KexAlgo, DigestAlgo, AeadAlgo>
+where
+    KexAlgo: Kex,
+    for<'privkey> <KexAlgo as Kex>::Public: From<&'privkey <KexAlgo as Kex>::EphemeralPrivate>,
+    DigestAlgo: Digest + Input + FixedOutput + Default + Clone + BlockInput + Reset,
+    AeadAlgo: Aead + NewAead,
+{
+    _kex: PhantomData<fn() -> KexAlgo>,
+    _digest: PhantomData<fn() -> DigestAlgo>,
+    _aead: PhantomData<fn() -> AeadAlgo>,
+}
+
+impl<KexAlgo, DigestAlgo, AeadAlgo> CryptoBox<KexAlgo> for HkdfBox<KexAlgo, DigestAlgo, AeadAlgo>
+where
+    KexAlgo: Kex,
+    for<'privkey> <KexAlgo as Kex>::Public: From<&'privkey <KexAlgo as Kex>::EphemeralPrivate>,
+    DigestAlgo: Digest + Input + FixedOutput + Default + Clone + BlockInput + Reset,
+    AeadAlgo: Aead + NewAead,
+    // Note: I think all of these bounds should go away after RFC 2089 is implemented
+    // https://github.com/rust-lang/rfcs/blob/master/text/2089-implied-bounds.md
+    <<KexAlgo as Kex>::Public as ReprBytes>::Size:
+        ArrayLength<u8> + Unsigned + Add<AeadAlgo::TagSize>,
+    Sum<<KexAlgo::Public as ReprBytes>::Size, AeadAlgo::TagSize>: ArrayLength<u8>,
+    GenericArray<u8, <<KexAlgo as Kex>::Public as ReprBytes>::Size>: Concat<
+        u8,
+        AeadAlgo::TagSize,
+        Rest = GenericArray<u8, AeadAlgo::TagSize>,
+        Output = GenericArray<
+            u8,
+            <<<KexAlgo as Kex>::Public as ReprBytes>::Size as Add<AeadAlgo::TagSize>>::Output,
+        >,
+    >,
```

Comment:
This isn't a valid constraint for an Aead.

Reply:
The compiler tells me this is needed to make `curve_point_bytes.concat(mac)` compile,
I think all I'm telling it is that `KexAlgo::Public::Size + AeadAlgo::TagSize` is a number, and the size of a valid generic array.
I'm not sure though, why this is needed at this revision and not the previous revision.